builtin/rebase.c

```diff
@@ -0,0 +1,687 @@
+		ret = error(_("could not write index"));
+	free((void *)desc.buffer);
+
+	if (ret)
+		return ret;
+
+	reflog_action = getenv(GIT_REFLOG_ACTION_ENVIRONMENT);
+	strbuf_addf(&msg, "%s: ", reflog_action ? reflog_action : "rebase");
+	prefix_len = msg.len;
+
+	if (!get_oid("ORIG_HEAD", &oid_old_orig))
+		old_orig = &oid_old_orig;
+	if (!get_oid("HEAD", &oid_orig)) {
+		orig = &oid_orig;
+		strbuf_addstr(&msg, "updating ORIG_HEAD");
+		update_ref(msg.buf, "ORIG_HEAD", orig, old_orig, 0,
+			   UPDATE_REFS_MSG_ON_ERR);
+	} else if (old_orig)
+		delete_ref(NULL, "ORIG_HEAD", old_orig, 0);
+	strbuf_setlen(&msg, prefix_len);
+	strbuf_addstr(&msg, "updating HEAD");
+	ret = update_ref(msg.buf, "HEAD", oid, orig, 0, UPDATE_REFS_MSG_ON_ERR);
+
+	strbuf_release(&msg);
+
+	return ret;
+}
+
+int cmd_rebase(int argc, const char **argv, const char *prefix)
+{
+	enum {
+		NO_ACTION,
+		ACTION_CONTINUE,
+		ACTION_SKIP,
+		ACTION_ABORT,
+		ACTION_QUIT,
+		ACTION_EDIT_TODO,
+		ACTION_SHOW_CURRENT_PATCH
+	} action = NO_ACTION;
+	struct rebase_options options = { REBASE_UNSPECIFIED };
+	const char *branch_name = NULL;
+	int ret, flags, quiet = 0, total_argc = 0, in_progress = 0;
+	int ok_to_skip_pre_rebase = 1;
+	int keep_empty = 1;
+	int autosquash = 0;
+	int fork_point = 0;
+	int do_merge = 1;
+	struct strbuf msg = STRBUF_INIT;
+	struct strbuf revisions = STRBUF_INIT;
+	struct strbuf git_am_opt = STRBUF_INIT;
+	struct strbuf buf = STRBUF_INIT;
+	const char *restrict_revision = NULL;
+	struct object_id merge_base;
+	struct option builtin_rebase_options[] = {
+		OPT_STRING(0, "onto", &options.onto_name,
+			N_("revision"),
+			N_("rebase onto given branch instead of upstream")),
+		OPT_BOOL(0, "no-verify", &ok_to_skip_pre_rebase,
+			N_("allow pre-rebase hook to run")),
+		OPT_NEGBIT('q', "quiet", &options.flags,
+			N_("be quiet. implies --no-stat"),
+			REBASE_VERBOSE | REBASE_DIFFSTAT),
+		OPT_BIT('v', "verbose", &options.flags,
+			N_("display a diffstat of what changed upstream"),
+			REBASE_VERBOSE | REBASE_DIFFSTAT),
+		OPT_CMDMODE(0, "continue", &action, N_("continue"),
+			ACTION_CONTINUE),
+		OPT_CMDMODE(0, "abort", &action,
+			    N_("abort and check out the original branch"),
+			    ACTION_ABORT),
+		OPT_CMDMODE(0, "skip", &action,
+			    N_("skip current patch and continue"), ACTION_SKIP),
+		OPT_CMDMODE(0, "edit-todo", &action, N_("edit the todo list "
+			    "during an interactive rebase"), ACTION_EDIT_TODO),
+		OPT_CMDMODE(0, "quit", &action,
+			    N_("abort but keep HEAD where it is"), ACTION_QUIT),
+		OPT_CMDMODE(0, "show-current-patch", &action,
+			    N_("show the patch file being applied or merged"),
+			    ACTION_SHOW_CURRENT_PATCH),
+		OPT_BOOL(0, "no-keep-empty", &keep_empty,
+			N_("preserve empty commits during rebase")),
+		OPT_BOOL(0, "no-autosquash", &autosquash,
+			N_("move commits that begin with "
+			    "squash!/fixup! under -i")),
+		OPT_BOOL(0, "no-fork-point", &fork_point,
```

Comment:
So now, the option `--fork-point` is parsed.

It is parsed incorrectly, though, as the "no-" prefix inverts the meaning: it would set `fork_point` to `1` in the *wrong* case, i.e. when the user called `rebase` with the `--no-fork-point` option.

However, even then, it only makes sense if you convert the shell script code that actually *acts* on `fork_point`:

```sh
if test "$fork_point" = t
then
	new_upstream=$(git merge-base --fork-point "$upstream_name" \
			"${switch_to:-HEAD}")
	if test -n "$new_upstream"
	then
		restrict_revision=$new_upstream
	fi
fi
```

When you do that, keep in mind that the C equivalent of `if test "fork_point" = t` is *not* `if (fork_point)`, as it really wants to handle *only* the `true` case, and not the `auto` case. So you will have to use either `if (fork_point == 1)` or `if (fork_point > 0)` (personally, I would much prefer the latter).

To convert the actual call to `git merge-base --fork-point`, you will have to inspect `builtin/merge-base.c` and see how you can replicate the functionality in C. From a cursory look, it appears to me that the entire logic -- apart from determining the merge bases -- is implemented in a file-local (`static`) function in `builtin/merge-base.c`, and that this probably needs to be refactored (my preference would be to extract the part after parsing the `oid`s from the two arguments into a function that lives right after the `merge_bases_many()` function in `commit.h` and `commit.c`).

But this is not a trivial task. Maybe finish the basic `REBASE_AM` stuff first?